src/QsCompiler/Transformations/CallGraph/CallGraphWalker.cs

```diff
@@ -0,0 +1,507 @@
+            {
+                internal TNode CurrentNode;
+                internal readonly TGraph Graph;
+
+                // The type parameter resolutions of the current expression.
+                internal IEnumerable<TypeParameterResolutions> ExpTypeParamResolutions = new List<TypeParameterResolutions>();
+                internal QsNullable<Position> CurrentStatementOffset;
+                internal QsNullable<Range> CurrentExpressionRange;
+                internal readonly Stack<TNode> RequestStack = new Stack<TNode>(); // Used to keep track of the nodes that still need to be walked by the walker.
+                internal readonly HashSet<TNode> ResolvedNodeSet = new HashSet<TNode>(); // Used to keep track of the nodes that have already been walked by the walker.
+
+                internal TransformationState(TGraph graph)
+                {
+                    this.Graph = graph;
+                }
+
+                /// <summary>
+                /// Adds dependency to the graph from the current callable to the callable referenced by the given identifier.
+                /// </summary>
+                internal abstract void AddDependency(QsQualifiedName identifier);
+            }
+
+            public class StatementWalker<TState> : StatementTransformation<TState>
+                where TState : TransformationState
+            {
+                public StatementWalker(SyntaxTreeTransformation<TState> parent) : base(parent, TransformationOptions.NoRebuild)
+                {
+                }
+
+                public override QsStatement OnStatement(QsStatement stm)
+                {
+                    this.SharedState.CurrentStatementOffset = stm.Location.IsValue
+                        ? QsNullable<Position>.NewValue(stm.Location.Item.Offset)
+                        : QsNullable<Position>.Null;
+                    return base.OnStatement(stm);
+                }
+            }
+
+            public class ExpressionWalker<TState> : ExpressionTransformation<TState>
+                where TState : TransformationState
+            {
+                public ExpressionWalker(SyntaxTreeTransformation<TState> parent) : base(parent, TransformationOptions.NoRebuild)
+                {
+                }
+
+                public override TypedExpression OnTypedExpression(TypedExpression ex)
+                {
+                    var contextRange = this.SharedState.CurrentExpressionRange;
+                    this.SharedState.CurrentExpressionRange = ex.Range;
+
+                    if (ex.TypeParameterResolutions.Any())
+                    {
+                        this.SharedState.ExpTypeParamResolutions = this.SharedState.ExpTypeParamResolutions.Prepend(ex.TypeParameterResolutions);
+                    }
+                    var rtrn = base.OnTypedExpression(ex);
+
+                    this.SharedState.CurrentExpressionRange = contextRange;
+
+                    return rtrn;
```

Comment:
For consistency with the other transformations below:

```suggestion
                    }

                    var result = base.OnTypedExpression(ex);
                    this.SharedState.CurrentExpressionRange = contextRange;
                    return result;
```

Reply:
done